File: z_lc/5-longest-palindromic-substring/solution-3-manachers/longest-palindromic-substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        #include <algorithm>
        
        const char SEPARATOR = '.';
        string S(1, SEPARATOR);
        for(auto c : s) {
            S.push_back(c);
            S.push_back(SEPARATOR);
        }
        int n = S.size();
        
        vector<int> lens(n);
        int c = 0, r = 0;
        int center = 0, length = 0;
        for (int i = 0; i < n; ++i) {
            if (i < r) {
                int i_prime = c - (i - c);
                lens[i] = std::min(r - i, lens[i_prime]);
            }
            
            int x, y;
            while ((x=i-1-lens[i]) >= 0 && (y=i+1+lens[i]) < n && S[x]==S[y]) {
                ++lens[i];
            }
            
            if (i + lens[i] > r) {
                r = i + lens[i];
                c = i;
            }
            
            if (lens[i] > length) {
                length = lens[i];
                center = i;
            }
        }
        int start = (center - length) / 2;
        return s.substr(start, length);
    }
};

File: z_lc/5-longest-palindromic-substring/solution-1-dp/longest-palindromic-substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        n = len(s)
        
        memo = [[False]*n for i in range(n)]
        
        start, length = 0, 0
        for i in range(n-1, -1, -1):
            for j in range(i, n):
                if i == j:
                    memo[i][j] = True
                elif i+1 == j:
                    memo[i][j] = (s[i] == s[j])
                else:
                    memo[i][j] = s[i] == s[j] and memo[i+1][j-1]
                if memo[i][j] and length < j - i + 1:
                    start, length = i, j - i + 1
        return s[start : start+length]